network-store-iidm-impl/src/main/java/com/powsybl/network/store/iidm/impl/util/JGraphTGraph.java

```diff
@@ -0,0 +1,91 @@
+ */
+public class JGraphTGraph {
+
+    private final Graph<Integer, Edge> graph;
+
+    public JGraphTGraph(Graph<Integer, Edge> graph) {
+        this.graph = graph;
+    }
+
+    public List<List<Edge>> findAllPaths(int from,
+                                          Predicate<Integer> pathComplete,
+                                          Predicate<Edge> pathCancelled,
+                                          Comparator<List<Edge>> comparator) {
+
+        List<List<Edge>> paths = new ArrayList<>();
+        BitSet encountered = new BitSet(Collections.max(graph.vertexSet()) + 1);
+        List<Edge> path = new ArrayList<>();
+        findAllPaths(from, pathComplete, pathCancelled, path, encountered, paths);
+
+        // sort paths by size according to the given comparator
+        paths.sort(comparator);
+        return paths;
+    }
+
+    private void findAllPaths(int v, Predicate<Integer> pathComplete, Predicate<Edge> pathCancelled,
+                              List<Edge> path, BitSet encountered, List<List<Edge>> paths) {
+        if (v < 0) {
+            throw new PowsyblException("Invalid vertex " + v);
+        }
+        encountered.set(v, true);
+        List<Edge> adjacentEdges = graph.edgesOf(v).stream().toList();
+        for (int i = 0; i < adjacentEdges.size(); i++) {
+            Edge edge = adjacentEdges.get(i);
+            if (pathCancelled != null && pathCancelled.test(edge)) {
+                // Do not continue on this path if the edge cannot be traversed
+                continue;
+            }
+            int v1 = graph.getEdgeSource(edge);
+            int v2 = graph.getEdgeTarget(edge);
+            List<Edge> path2;
+            BitSet encountered2;
+            if (i < adjacentEdges.size() - 1) {
+                path2 = new ArrayList<>(path);
+                encountered2 = new BitSet(graph.vertexSet().size());
+                encountered2.or(encountered);
+            } else {
+                path2 = path;
+                encountered2 = encountered;
+            }
+            if (v == v2) {
+                findAllPaths(edge, v1, pathComplete, pathCancelled, path2, encountered2, paths);
+            } else if (v == v1) {
+                findAllPaths(edge, v2, pathComplete, pathCancelled, path2, encountered2, paths);
+            } else {
+                throw new IllegalStateException();
+            }
+        }
+    }
+
+    private void findAllPaths(Edge edge, int v1or2, Predicate<Integer> pathComplete, Predicate<Edge> pathCancelled,
+                                 List<Edge> path, BitSet encountered, List<List<Edge>> paths) {
+        if (encountered.get(v1or2)) {
+            return;
+        }
+        path.add(edge);
+        if (Boolean.TRUE.equals(pathComplete.test(v1or2))) {
+            paths.add(path);
+        } else {
+            findAllPaths(v1or2, pathComplete, pathCancelled, path, encountered, paths);
+        }
+    }
```

Comment:
Can't we have a simple DFS like we already have in traverseFromNodeDFS() or maybe I'm missing something?
```java
    private void findAllPaths(int currentVertex,
                              Predicate<Integer> pathComplete,
                              Predicate<Edge> pathCancelled,
                              List<Edge> currentPath,
                              BitSet encountered,
                              List<List<Edge>> paths) {
        if (currentVertex < 0) {
            throw new PowsyblException("Invalid vertex " + currentVertex);
        }
        if (encountered.get(currentVertex)) {
            return;
        }
        encountered.set(currentVertex, true);

        if (pathComplete.test(currentVertex)) {
            paths.add(new ArrayList<>(currentPath));
        } else {
            for (Edge adjacentEdge : graph.edgesOf(currentVertex)) {
                if (pathCancelled != null && pathCancelled.test(adjacentEdge)) {
                    // Do not continue on this path if the edge cannot be traversed
                    continue;
                }
                int nextVertex = graph.getEdgeSource(adjacentEdge).equals(currentVertex) ? graph.getEdgeTarget(adjacentEdge) : graph.getEdgeSource(adjacentEdge);
                currentPath.add(adjacentEdge);
                findAllPaths(nextVertex, pathComplete, pathCancelled, currentPath, encountered, paths);
                // Remove the current edge from the path to backtrack
                currentPath.remove(adjacentEdge);
            }
        }
        // Allow currentVertex to be used in other paths
        encountered.clear(currentVertex);
    }
```

Reply:
A DFS could also be used here be I prefered to use the exact same algorithm as in powsybl-core.

Reply:
I was not aware it was a duplicate from core. We should try to use the core methods at some point instead of duplicating. I'm not sure what prevents us from that?

Reply:
Right now we are using jgrapht graph implementation in Network-store while we are using a custom implementation in core, so we can't use the methods from core right now.
In the future, we could try to find the equivalent methods in JGraphT in order to switch the implementation in core, or we could switch all uses of JGraphT in network-store to use core custom implementation.